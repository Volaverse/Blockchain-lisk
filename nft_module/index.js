const { BaseModule } = require("lisk-sdk");
const { getAllNFTTokensAsJSON } = require("./nft");

const CreateNFTAsset = require("./transactions/create_nft_asset");
const PurchaseNFTAsset = require("./transactions/purchase_nft_asset");
const TransferNFTAsset = require("./transactions/transfer_nft_asset");
const TransferTestNFTAsset = require("./transactions/transfer_test_nft_asset");
const SellNFTAsset = require("./transactions/sell_nft_asset");
const EditNFTNameAsset = require("./transactions/edit_title_nft_asset")
const EditNFTDescAsset = require("./transactions/edit_description_nft_asset")
// Extend base module to implement your custom module
class NFTModule extends BaseModule {
  name = "nft";
  id = 1024;
  accountSchema = {
    type: "object",
    required: ["ownNFTs"],
    properties: {
      ownNFTs: {
        type: "array",
        fieldNumber: 1,
        items: {
          dataType: "bytes",
        },
      },
    },
    default: {
      ownNFTs: [],
    },
  };
  transactionAssets = [new CreateNFTAsset(), new PurchaseNFTAsset(), new TransferNFTAsset(),new TransferTestNFTAsset(),new SellNFTAsset(),new EditNFTNameAsset(),new EditNFTDescAsset()];
  actions = {
    // get all the registered NFT tokens from blockchain
    getAllNFTTokens: async () => getAllNFTTokensAsJSON(this._dataAccess),
  };
}

module.exports = { NFTModule };
