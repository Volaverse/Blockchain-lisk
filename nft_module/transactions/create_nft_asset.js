const { BaseAsset } = require("lisk-sdk");
const {
  getAllNFTTokens,
  setAllNFTTokens,
  createNFTToken,
} = require("../nft");

// 1.extend base asset to implement your custom asset
class CreateNFTAsset extends BaseAsset {
  // 2.define unique asset name and id
  name = "createNFT";
  id = 0;
  // 3.define asset schema for serialization
  schema = {
    $id: "lisk/nft/create",
    type: "object",
    required: ["minPurchaseMargin", "initValue", "name","category","imageUrl","x","y","area","landmark","type","bodypart","gender","serialNo"],
    properties: {
      minPurchaseMargin: {
        dataType: "uint32",
        fieldNumber: 1,
      },
      initValue: {
        dataType: "uint64",
        fieldNumber: 2,
      },
      name: {
        dataType: "string",
        fieldNumber: 3,
      },
      description: {
        dataType: "string",
        fieldNumber: 4,
      },
      category: {
        dataType: "uint32",
        fieldNumber: 5,
      },
      imageUrl: {
        dataType: "string",
        fieldNumber: 6,
      },
      x: {
        dataType: "string",
        fieldNumber: 7,
      },
      y: {
        dataType: "string",
        fieldNumber: 8,
      },
      threeDUrl: {
        dataType: "string",
        fieldNumber: 9,
      },
      area: {
        dataType: "string",
        fieldNumber: 10,
      },
      landmark: {
        dataType: "string",
        fieldNumber: 11,
      },
      type: {
        dataType: "string",
        fieldNumber: 12,
      },
      bodypart: {
        dataType: "string",
        fieldNumber: 13,
      },
      gender: {
        dataType: "string",
        fieldNumber: 14,
      },
      serialNo: {
        dataType: "string",
        fieldNumber: 15,
      },
    },
  };
  validate({asset}) {
    if (asset.initValue <= 0) {
      throw new Error("NFT init value is too low.");
    } else if (asset.minPurchaseMargin < 0 || asset.minPurchaseMargin > 100) {
      throw new Error("The NFT minimum purchase value needs to be between 0 and 100.");
    }

  };
  async apply({ asset, stateStore, reducerHandler, transaction }) {
    // 4.verify if sender has enough balance
    const allTokens = await getAllNFTTokens(stateStore);
    const senderAddress = transaction.senderAddress;
    const senderAccount = await stateStore.account.get(senderAddress);
    var dup=[];
    if(allTokens.length>0){
       dup = allTokens.filter((item)=> item.name==asset.name)
    }

    console.log("check for dup length "+dup.length)

    if(dup.length>0){
      throw new Error("Name of nft should be unique");
    }

    if (senderAddress.toString("hex") !='16c70194f16fa137d96168823f695d2ddb232554' && !(asset.category==3 || asset.category=='3')) {
      throw new Error("NFT cannot be created from this account"+senderAddress.toString("hex"));
    }
    // 5.create nft
    const nftToken = createNFTToken({
      name: asset.name,
      description: asset.description,
      ownerAddress: senderAddress,
      nonce: transaction.nonce,
      value: asset.initValue,
      minPurchaseMargin: asset.minPurchaseMargin,
      category: asset.category,
      imageUrl: asset.imageUrl,
      x:asset.x,
      y:asset.y,
      threeDUrl:asset.threeDUrl,
      area: asset.area,
      landmark:asset.landmark,
      type:asset.type,
      bodypart:asset.bodypart,
      gender:asset.gender,
      serialNo:asset.serialNo

    });

    // 6.update sender account with unique nft id
    senderAccount.nft.ownNFTs.push(nftToken.id);
    await stateStore.account.set(senderAddress, senderAccount);

    // 7.debit tokens from sender account to create nft
    await reducerHandler.invoke("token:debit", {
      address: senderAddress,
      amount: asset.initValue,
    });

    // 8.save nfts
    
    allTokens.push(nftToken);
    await setAllNFTTokens(stateStore, allTokens);
  }
}

module.exports = CreateNFTAsset;
