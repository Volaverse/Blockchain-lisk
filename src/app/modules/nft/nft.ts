/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import { codec, cryptography } from 'lisk-sdk';

export interface nftInterface {
	id: {
		dataType: 'bytes';
	};
	value: {
		dataType: 'uint64';
	};
	ownerAddress: {
		dataType: 'bytes';
	};
	minPurchaseMargin: {
		dataType: 'uint32';
	};
	name: {
		dataType: 'string';
	};
	description: {
		dataType: 'string';
	};
	category: {
		dataType: 'uint32';
	};
}
interface RegisteredNFT {
	registeredNFTTokens: nftInterface[];
}

export const registeredNFTTokensSchema = {
	$id: 'lisk/nft/registeredTokens',
	type: 'object',
	required: ['registeredNFTTokens'],
	properties: {
		registeredNFTTokens: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['id', 'value', 'ownerAddress', 'minPurchaseMargin', 'name', 'category'],
				properties: {
					id: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					value: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
					ownerAddress: {
						dataType: 'bytes',
						fieldNumber: 3,
					},
					minPurchaseMargin: {
						dataType: 'uint32',
						fieldNumber: 4,
					},
					name: {
						dataType: 'string',
						fieldNumber: 5,
					},
					description: {
						dataType: 'string',
						fieldNumber: 6,
					},
					category: {
						dataType: 'uint32',
						fieldNumber: 7,
					},
				},
			},
		},
	},
};

export const CHAIN_STATE_NFT_TOKENS = 'nft:registeredNFTTokens';

// returns nft object with id
export const createNFTToken = ({
	name,
	description,
	ownerAddress,
	nonce,
	value,
	minPurchaseMargin,
	category,
}) => {
	const nonceBuffer = Buffer.alloc(8);
	nonceBuffer.writeBigInt64LE(nonce);
	const seed = Buffer.concat([ownerAddress, nonceBuffer]);
	const id = cryptography.hash(seed);

	return {
		id,
		minPurchaseMargin,
		name,
		description,
		ownerAddress,
		value,
		category,
	};
};
// get list of all nft tokens
export const getAllNFTTokens = async stateStore => {
	const registeredTokensBuffer = await stateStore.chain.get(CHAIN_STATE_NFT_TOKENS);
	if (!registeredTokensBuffer) {
		return [];
	}

	const registeredTokens: any = codec.decode(registeredNFTTokensSchema, registeredTokensBuffer);

	return registeredTokens.registeredNFTTokens;
};
// returns nft tokens as json
export const getAllNFTTokensAsJSON = async dataAccess => {
	const registeredTokensBuffer = await dataAccess.getChainState(CHAIN_STATE_NFT_TOKENS);

	if (!registeredTokensBuffer) {
		return [];
	}

	const registeredTokens = codec.decode<RegisteredNFT>(
		registeredNFTTokensSchema,
		registeredTokensBuffer,
	);

	return codec.toJSON<RegisteredNFT>(registeredNFTTokensSchema, registeredTokens)
		.registeredNFTTokens;
};
// set nft tokens to state store
export const setAllNFTTokens = async (stateStore, NFTTokens) => {
	const registeredTokens = {
		registeredNFTTokens: NFTTokens.sort((a, b) => a.id.compare(b.id)),
	};

	await stateStore.chain.set(
		CHAIN_STATE_NFT_TOKENS,
		codec.encode(registeredNFTTokensSchema, registeredTokens),
	);
};

module.exports = {
	registeredNFTTokensSchema,
	CHAIN_STATE_NFT_TOKENS,
	getAllNFTTokens,
	setAllNFTTokens,
	getAllNFTTokensAsJSON,
	createNFTToken,
};
