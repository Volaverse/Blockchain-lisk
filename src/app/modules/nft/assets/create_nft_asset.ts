/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

import { BaseAsset } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens, createNFTToken } from '../nft';

export class CreateNftAsset extends BaseAsset {
	public name = 'createNft';
	public id = 0;

	// define schema for asset
	public schema = {
		$id: 'nft/createNft-asset',
		title: 'CreateNftAsset transaction asset for nft module',
		type: 'object',
		required: ['minPurchaseMargin', 'initValue', 'name', 'description', 'category'],
		properties: {
			minPurchaseMargin: {
				dataType: 'uint32',
				fieldNumber: 1,
			},
			initValue: {
				dataType: 'uint64',
				fieldNumber: 2,
			},
			name: {
				dataType: 'string',
				fieldNumber: 3,
			},
			description: {
				dataType: 'string',
				fieldNumber: 4,
			},
			category: {
				dataType: 'uint32',
				fieldNumber: 5,
			},
		},
	};

	public validate({ asset }): void {
		// validate your asset
		if (asset.initValue <= 0) {
			throw new Error('NFT init value is too low.');
		} else if (asset.minPurchaseMargin < 0 || asset.minPurchaseMargin > 100) {
			throw new Error('The NFT minimum purchase value needs to be between 0 and 100.');
		}
	}

	public async apply({ asset, transaction, stateStore, reducerHandler }): Promise<void> {
		const allTokens = await getAllNFTTokens(stateStore);
		const { senderAddress } = transaction;
		const senderAccount = await stateStore.account.get(senderAddress);
		let dup = [];
		// Checking for unique nft name
		if (allTokens.length > 0) {
			dup = allTokens.filter(item => item.name === asset.name);
		}

		if (dup.length > 0) {
			throw new Error('Name of nft should be unique');
		}
		// category 3 is for usernameNFT which anyone can create
		// Other categories NFT can only be created by a particular address
		if (
			senderAddress.toString('hex') !== '16c70194f16fa137d96168823f695d2ddb232554' &&
			!(asset.category === 3 || asset.category === '3')
		) {
			throw new Error(`NFT cannot be created from this account ${senderAddress.toString('hex')}`);
		}
		// create nft
		const nftToken = createNFTToken({
			name: asset.name,
			description: asset.description,
			ownerAddress: senderAddress,
			nonce: transaction.nonce,
			value: asset.initValue,
			minPurchaseMargin: asset.minPurchaseMargin,
			category: asset.category,
		});

		// 6.update sender account with unique nft id
		senderAccount.nft.ownNFTs.push(nftToken.id);
		await stateStore.account.set(senderAddress, senderAccount);

		// 7.debit tokens from sender account to create nft
		await reducerHandler.invoke('token:debit', {
			address: senderAddress,
			amount: asset.initValue,
		});

		// 8.save nfts

		allTokens.push(nftToken);
		await setAllNFTTokens(stateStore, allTokens);
	}
}
