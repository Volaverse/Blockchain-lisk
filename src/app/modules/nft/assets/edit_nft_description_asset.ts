/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens } from '../nft';

export class EditNftDescriptionAsset extends BaseAsset {
	public name = 'editNftDescription';
	public id = 4;

	// define schema for asset
	public schema = {
		$id: 'nft/editNftDescription-asset',
		title: 'EditNftDescriptionAsset transaction asset for nft module',
		type: 'object',
		required: ['nftId', 'description'],
		properties: {
			nftId: {
				dataType: 'bytes',
				fieldNumber: 1,
			},
			description: {
				dataType: 'string',
				fieldNumber: 2,
			},
		},
	};

	public validate({ asset }): void {
		if (asset.nftId === '' || !asset.nftId) {
			throw new Error('NFT id should have value');
		}
		if (asset.description === '' || !asset.description) {
			throw new Error('description cannot be empty');
		}
	}

	public async apply({ asset, transaction, stateStore }): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex(t => t.id.equals(asset.nftId));

		// verify if the nft exists
		if (nftTokenIndex < 0) {
			throw new Error('Token id not found');
		}

		const token = nftTokens[nftTokenIndex];
		const tokenOwnerAddress = token.ownerAddress;
		const { senderAddress } = transaction;
		// verify that the sender owns the nft

		if (!tokenOwnerAddress.equals(senderAddress)) {
			throw new Error('An NFT description can only be changed by the owner of the NFT.');
		}

		// setting the minimum sell value
		token.description = asset.description;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);
	}
}
