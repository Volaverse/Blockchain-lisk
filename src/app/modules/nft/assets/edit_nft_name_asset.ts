/* eslint-disable @typescript-eslint/restrict-plus-operands */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from "lisk-sdk";
import { getAllNFTTokens, setAllNFTTokens } from "../nft";

export class EditNftNameAsset extends BaseAsset {
  public name = "editNftName";
  public id = 3;

  // define schema for asset
  public schema = {
    $id: "nft/editNftName-asset",
    title: "EditNftNameAsset transaction asset for nft module",
    type: "object",
    required: ["nftId", "name"],
    properties: {
      nftId: {
        dataType: "bytes",
        fieldNumber: 1,
      },
      name: {
        dataType: "string",
        fieldNumber: 2,
      },
    },
  };

  public validate({ asset }): void {
    if (asset.nftId === "" || !asset.nftId) {
      throw new Error("NFT id should have value");
    }
    if (asset.name === "" || !asset.name) {
      throw new Error("description cannot be empty");
    }
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }): Promise<void> {
    const nftTokens = await getAllNFTTokens(stateStore);
    const nftTokenIndex = nftTokens.findIndex((t) => t.id.equals(asset.nftId));

    if (asset.name === "") {
      throw new Error("Name cannot be empty");
    }

    // verify if the nft exists
    if (nftTokenIndex < 0) {
      throw new Error("Token id not found");
    }
    // checking for unique nft name
    let dup = [];
    if (nftTokens.length > 0) {
      dup = nftTokens.filter((item) => item.name === asset.name);
    }

    if (dup.length > 0) {
      throw new Error("Name of nft should be unique");
    }
    const token = nftTokens[nftTokenIndex];
    if (token.category !== 3) {
      throw new Error("Name can only be changed of username category");
    }
    const tokenOwnerAddress = token.ownerAddress;
    const { senderAddress } = transaction;
    // verify that the sender owns the nft

    if (!tokenOwnerAddress.equals(senderAddress)) {
      throw new Error("An NFT can only be renamed by the owner of the NFT.");
    }

    // setting the minimum sell value
    token.name = asset.name;
    nftTokens[nftTokenIndex] = token;
    await setAllNFTTokens(stateStore, nftTokens);
  }
}
