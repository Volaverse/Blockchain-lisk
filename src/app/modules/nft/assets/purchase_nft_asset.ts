/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/restrict-plus-operands */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from 'lisk-sdk';

import { getAllNFTTokens, setAllNFTTokens } from '../nft';

export class PurchaseNftAsset extends BaseAsset {
	public name = 'purchaseNft';
	public id = 1;

	// define schema for asset
	public schema = {
		$id: 'nft/purchaseNft-asset',
		title: 'PurchaseNftAsset transaction asset for nft module',
		type: 'object',
		required: ['nftId', 'purchaseValue'],
		properties: {
			nftId: {
				dataType: 'bytes',
				fieldNumber: 1,
			},
			purchaseValue: {
				dataType: 'uint64',
				fieldNumber: 2,
			},
			name: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
	};

	public validate({ asset }): void {
		// validate your asset
		if (asset.nftId === '' || !asset.nftId) {
			throw new Error('NFT id should have value');
		}
		if (asset.purchaseValue === '' || !asset.purchaseValue) {
			throw new Error('purchaseValue should have value');
		}
		if (asset.purchaseValue <= 0) {
			throw new Error('The NFT purchase value needs to be more than 0');
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore, reducerHandler }): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex(t => t.id.equals(asset.nftId));

		// verify if purchasing nft exists
		if (nftTokenIndex < 0) {
			throw new Error('Token id not found');
		}
		const token = nftTokens[nftTokenIndex];
		const tokenOwner = await stateStore.account.get(token.ownerAddress);
		const tokenOwnerAddress = tokenOwner.address;

		// 5.verify if minimum nft purchasing condition met
		if (token && token.minPurchaseMargin === 0) {
			throw new Error('This NFT token can not be purchased');
		}

		const tokenCurrentValue = token.value;
		const tokenMinPurchaseValue =
			tokenCurrentValue + (tokenCurrentValue * BigInt(token.minPurchaseMargin)) / BigInt(100);
		const { purchaseValue } = asset;

		if (tokenMinPurchaseValue > purchaseValue) {
			throw new Error(
				`Token can not be purchased. Purchase value is too low. Minimum value: ${tokenMinPurchaseValue}`,
			);
		}

		const purchaserAddress = transaction.senderAddress;
		const purchaserAccount = await stateStore.account.get(purchaserAddress);

		// remove nft from owner account
		const ownerTokenIndex = tokenOwner.nft.ownNFTs.findIndex(a => a.equals(token.id));
		tokenOwner.nft.ownNFTs.splice(ownerTokenIndex, 1);
		await stateStore.account.set(tokenOwnerAddress, tokenOwner);

		// add nft to purchaser account
		purchaserAccount.nft.ownNFTs.push(token.id);
		await stateStore.account.set(purchaserAddress, purchaserAccount);

		token.ownerAddress = purchaserAddress;
		token.value = purchaseValue;
		token.minPurchaseMargin = 0;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);

		// debit LSK tokens from purchaser account
		await reducerHandler.invoke('token:debit', {
			address: purchaserAddress,
			amount: purchaseValue,
		});

		// credit LSK tokens to purchaser account
		await reducerHandler.invoke('token:credit', {
			address: tokenOwnerAddress,
			amount: purchaseValue,
		});
	}
}
