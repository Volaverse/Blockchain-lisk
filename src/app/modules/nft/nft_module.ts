/* eslint-disable class-methods-use-this */

import { BaseModule } from 'lisk-sdk';
import { CreateNftAsset } from './assets/create_nft_asset';
import { EditNftDescriptionAsset } from './assets/edit_nft_description_asset';
import { EditNftNameAsset } from './assets/edit_nft_name_asset';
import { PurchaseNftAsset } from './assets/purchase_nft_asset';
import { SellNftAsset } from './assets/sell_nft_asset';
import { getAllNFTTokensAsJSON } from './nft';

export class NftModule extends BaseModule {
	public actions = {
		getAllNFTTokens: async () => getAllNFTTokensAsJSON(this._dataAccess),
	};
	public reducers = {};
	public name = 'nft';
	public transactionAssets = [
		new CreateNftAsset(),
		new PurchaseNftAsset(),
		new SellNftAsset(),
		new EditNftNameAsset(),
		new EditNftDescriptionAsset(),
	];
	public events = [];
	public id = 1024;
	public accountSchema = {
		type: 'object',
		required: ['ownNFTs'],
		properties: {
			ownNFTs: {
				type: 'array',
				fieldNumber: 1,
				items: {
					dataType: 'bytes',
				},
			},
		},
		default: {
			ownNFTs: [],
		},
	};
}
