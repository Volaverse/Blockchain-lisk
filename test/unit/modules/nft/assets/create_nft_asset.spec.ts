/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable jest/expect-expect */
import { ReducerHandler, StateStore, testing } from "lisk-sdk";
import { CreateNftAsset } from "../../../../../src/app/modules/nft/assets/create_nft_asset";
import { NftModule } from "../../../../../src/app/modules/nft/nft_module";

interface accountInterface {
  nft: {
    ownNFTs: any[];
  };
}

describe("CreateNftAsset", () => {
  let transactionAsset: CreateNftAsset;

  beforeEach(() => {
    transactionAsset = new CreateNftAsset();
  });

  describe("constructor", () => {
    it("should have valid id", () => {
      expect(transactionAsset.id).toEqual(0);
    });

    it("should have valid name", () => {
      expect(transactionAsset.name).toEqual("createNft");
    });

    it("should have valid schema", () => {
      expect(transactionAsset.schema).toMatchSnapshot();
    });
  });

  describe("validate", () => {
    describe("schema validation", () => {
      it("should pass on validate asset", () => {
        const context = testing.createValidateAssetContext({
          asset: {
            name: "testing name ",
            description: "description is ",
            minPurchaseMargin: "1",
            initValue: "10",
            category: "3",
            imageUrl:"imageUrl",
            threeDUrl:"threeDUrl"
          },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });

        expect(() => transactionAsset.validate(context)).not.toThrow();
      });

      it("it should throw error that name is missing", () => {
        const context = testing.createValidateAssetContext({
          asset: {
            name: "test",
            description: "description is ",
            minPurchaseMargin: "1",
            initValue: "-1",
            category: "3",
          },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });

        expect(() => transactionAsset.validate(context)).toThrow(
          "NFT init value is too low."
        );
      });
    });
  });

  describe("apply", () => {
    let stateStore: StateStore;
    let reducerHandler: ReducerHandler;
    let account: any;
    let context;
    let contextInvalid;

    beforeEach(() => {
      account = testing.fixtures.createDefaultAccount([NftModule]);
      stateStore = new testing.mocks.StateStoreMock({ accounts: [account] });
      reducerHandler = testing.mocks.reducerHandlerMock;
      context = testing.createApplyAssetContext({
        stateStore,
        reducerHandler,
        asset: {
          name: "test 1",
          description: "description is ",
          minPurchaseMargin: "1",
          initValue: "1",
          category: "3",
          imageUrl:"imageUrl",
          threeDUrl:"threeDUrl"
        },
        transaction: {
          senderAddress: account.address,
          nonce: BigInt(0),
        } as any,
      });
      contextInvalid = testing.createApplyAssetContext({
        stateStore,
        reducerHandler,
        asset: {
          name: "test 1",
          description: "description is ",
          minPurchaseMargin: "1",
          initValue: "1",
          category: "2",
        },
        transaction: {
          senderAddress: account.address,
          nonce: BigInt(0),
        } as any,
      });

      jest.spyOn(stateStore.chain, "get");
      jest.spyOn(stateStore.chain, "set");
      jest.spyOn(reducerHandler, "invoke");
    });

    describe("valid cases", () => {
      it("should create new nft", async () => {
        await transactionAsset.apply(context);
        const updatedSender: accountInterface = await stateStore.account.get(
          account.address
        );

        expect(updatedSender.nft.ownNFTs).toHaveLength(1);
      });
    });

    describe("invalid cases", () => {
      it("should not create new nft", async () => {
        await expect(transactionAsset.apply(contextInvalid)).rejects.toThrow();
      });
    });
  });
});
