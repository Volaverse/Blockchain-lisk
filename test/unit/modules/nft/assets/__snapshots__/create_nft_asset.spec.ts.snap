// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNftAsset constructor should have valid schema 1`] = `
Object {
  "$id": "nft/createNft-asset",
  "properties": Object {
    "category": Object {
      "dataType": "uint32",
      "fieldNumber": 5,
    },
    "description": Object {
      "dataType": "string",
      "fieldNumber": 4,
    },
    "imageUrl": Object {
      "dataType": "string",
      "fieldNumber": 6,
    },
    "initValue": Object {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "minPurchaseMargin": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
    "name": Object {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "threeDUrl": Object {
      "dataType": "string",
      "fieldNumber": 7,
    },
  },
  "required": Array [
    "minPurchaseMargin",
    "initValue",
    "name",
    "description",
    "category",
  ],
  "title": "CreateNftAsset transaction asset for nft module",
  "type": "object",
}
`;
